### PR TITLE
Verify all available tools are downloadable

### LINKED ISSUE #557: Verify all available tools are downloadable
## Expected Behaviour

`arkade` CI should be aware and fail once a tool isn't available any longer.

## Current Behaviour

Currently there is no way to know that a tool got broken unless someone tries to use the tool and fail.

There are unit tests that verify that the outcome for getting a tool URL is as expected on https://github.com/alexellis/arkade/blob/master/pkg/get/get_test.go. 
This method has couple of disadvantages:
1. It requires to manually add a new tool test with every available OS / arch.
- Consumes a lot of time.
- Some combinations get neglected.
2. While there is no standard for installations URLs, tools might change and break the expected URL template as `arkade` currently supports. In that case, the UT will pass until someone will notify that the URL isn't actually available.

## Are you a GitHub Sponsor (Yes/No?)

Check at https://github.com/sponsors/alexellis

- [ ] Yes
- [x] No

## Possible Solution

* Extend the current test coverage

1. Add a system test to cover all the available tools with a simple `arkade get <tool>`. 
**Pro** - It would cover an e2e test and will assure that an installation works; Can measure installation timings.
**Con** - This approach will consume more resources and take more time. Rate limiting / token consumption might be affected.
In that case, maybe this test shouldn't run for every PR & locally, but only on cron jobs (further on the 2nd bullet).

<details> <summary> Code example (click to expand) </summary>

```golang
func Test_GetTools(t *testing.T) {
	tools := get.MakeTools()
	for _, tool := range tools {
		t.Run(tool.Name, func(t *testing.T) {
			cmd := MakeGet()
			cmd.SetArgs([]string{"--stash=false", tool.Name})
			err := cmd.Execute()
			if err != nil {
				t.Fatal(err)
			}
		})
	}
}
```

Output:
```text
--- PASS: Test_GetTools (322.87s)
    --- PASS: Test_GetTools/faas-cli (3.21s)
    --- PASS: Test_GetTools/helm (3.21s)
    --- PASS: Test_GetTools/helmfile (5.06s)
    --- PASS: Test_GetTools/jq (1.26s)
    --- PASS: Test_GetTools/kubectl (3.81s)
    --- PASS: Test_GetTools/kubectx (0.38s)
    --- PASS: Test_GetTools/kubens (0.39s)
    --- PASS: Test_GetTools/kind (1.42s)
    --- PASS: Test_GetTools/k3s (6.93s)
    --- PASS: Test_GetTools/autok3s (2.63s)
    --- PASS: Test_GetTools/devspace (5.87s)
    --- PASS: Test_GetTools/tilt (4.29s)
    --- PASS: Test_GetTools/k3d (2.51s)
    --- PASS: Test_GetTools/k3sup (1.42s)
    --- PASS: Test_GetTools/arkade (1.69s)
    --- PASS: Test_GetTools/kubeseal (3.67s)
    --- PASS: Test_GetTools/inletsctl (3.24s)
    --- PASS: Test_GetTools/osm (3.64s)
    --- PASS: Test_GetTools/linkerd2 (4.00s)
    --- PASS: Test_GetTools/kubebuilder (3.14s)
    --- PASS: Test_GetTools/kustomize (1.63s)
    --- PASS: Test_GetTools/doctl (2.19s)
    --- PASS: Test_GetTools/k9s (3.50s)
    --- PASS: Test_GetTools/popeye (2.97s)
    --- PASS: Test_GetTools/civo (1.89s)
    --- PASS: Test_GetTools/terraform (4.76s)
    --- PASS: Test_GetTools/vagrant (1.00s)
    --- PASS: Test_GetTools/packer (4.02s)
    --- PASS: Test_GetTools/gh (2.48s)
    --- PASS: Test_GetTools/pack (2.25s)
    --- PASS: Test_GetTools/buildx (6.24s)
    --- PASS: Test_GetTools/kops (15.57s)
    --- PASS: Test_GetTools/krew (1.93s)
    --- PASS: Test_GetTools/minikube (7.16s)
    --- PASS: Test_GetTools/stern (2.68s)
    --- PASS: Test_GetTools/kail (2.28s)
    --- PASS: Test_GetTools/yq (1.63s)
    --- PASS: Test_GetTools/kube-bench (2.40s)
    --- PASS: Test_GetTools/hugo (3.39s)
    --- PASS: Test_GetTools/docker-compose (1.40s)
    --- PASS: Test_GetTools/opa (5.39s)
    --- PASS: Test_GetTools/mc (50.18s)
    --- PASS: Test_GetTools/nats (4.12s)
    --- PASS: Test_GetTools/argocd (10.97s)
    --- PASS: Test_GetTools/nerdctl (2.14s)
    --- PASS: Test_GetTools/istioctl (4.91s)
    --- PASS: Test_GetTools/tkn (4.15s)
    --- PASS: Test_GetTools/inlets-pro (2.46s)
    --- PASS: Test_GetTools/kim (4.75s)
    --- PASS: Test_GetTools/trivy (2.97s)
    --- PASS: Test_GetTools/flux (3.50s)
    --- PASS: Test_GetTools/polaris (2.81s)
    --- PASS: Test_GetTools/influx (1.28s)
    --- PASS: Test_GetTools/argocd-autopilot (7.86s)
    --- PASS: Test_GetTools/nova (3.01s)
    --- PASS: Test_GetTools/kubetail (0.43s)
    --- PASS: Test_GetTools/kgctl (5.38s)
    --- PASS: Test_GetTools/porter (5.70s)
    --- PASS: Test_GetTools/k0s (18.34s)
    --- PASS: Test_GetTools/k0sctl (2.44s)
    --- PASS: Test_GetTools/metal (1.56s)
    --- PASS: Test_GetTools/kanctl (8.73s)
    --- PASS: Test_GetTools/kubestr (2.10s)
    --- PASS: Test_GetTools/k10multicluster (5.27s)
    --- PASS: Test_GetTools/k10tools (9.22s)
    --- PASS: Test_GetTools/cosign (7.09s)
    --- PASS: Test_GetTools/rekor-cli (2.85s)
    --- PASS: Test_GetTools/tfsec (1.69s)
    --- PASS: Test_GetTools/dive (1.06s)
    --- PASS: Test_GetTools/goreleaser (2.80s)
    --- PASS: Test_GetTools/kubescape (4.57s)
```

</details>

2. Add UT that goes over all the available tools and checks that the installation URL is accessible.
**Pro**: Much faster test.
**Con**: Will cover only that the URL is available. The installation might not yet be successful (there might be more parts that get broken, for example the `decompress` method)

* Add a daily [GHA cron job](https://docs.github.com/en/actions/learn-github-actions/events-that-trigger-workflows#schedule) that runs all the tests. 
Extra: In case of a failure, [send a Slack message](https://github.com/marketplace/actions/slack-send)

## Context

Having more coverage will cause faster identification of failures.
- https://github.com/alexellis/arkade/issues/550
- https://github.com/alexellis/arkade/issues/413
- https://github.com/alexellis/arkade/issues/342

## Your Environment

* What Kubernetes distribution are you using?

```
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.3", GitCommit:"2e7996e3e2712684bc73f0dec0200d64eec7fe40", GitTreeState:"clean", BuildDate:"2020-05-20T12:52:00Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.1", GitCommit:"5e58841cce77d4bc13713ad2b91fa0d961e69192", GitTreeState:"clean", BuildDate:"2021-05-21T23:01:33Z", GoVersion:"go1.16.4", Compiler:"gc", Platform:"linux/amd64"}
```

* Operating System and version (e.g. Linux, Windows, MacOS):

```
NAME="Ubuntu"
VERSION="21.04 (Hirsute Hippo)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 21.04"
VERSION_ID="21.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=hirsute
UBUNTU_CODENAME=hirsute
```

* What arkade version is this?

```
            _             _      
  __ _ _ __| | ____ _  __| | ___ 
 / _` | '__| |/ / _` |/ _` |/ _ \
| (_| | |  |   < (_| | (_| |  __/
 \__,_|_|  |_|\_\__,_|\__,_|\___|

Get Kubernetes apps the easy way

Version: 0.8.2
Git Commit: 8b3fbc0994129f6c14daef467593ae02799409c0
```

### PR DESCRIPTION
**Pull request by yankeexe**
Closes #557 

## Description

Add a new ci package that can be run independently from the CI to check for the tools whose URL is failing. 

## Motivation and Context

- [x] I have raised an issue to propose this change, which has been given a label of `design/approved` by a maintainer ([required](https://github.com/alexellis/arkade/blob/master/CONTRIBUTING.md))

## How Has This Been Tested?

This has been tested on my fork, here's the preview of the job run. 
https://github.com/yankeexe/arkade/runs/5259401645?check_suite_focus=true
The whole job run takes less than 1 minute to complete.

## Are you a GitHub Sponsor (Yes/No?)

- [ ] Yes
- [x] No

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Documentation

- [x] I have updated the list of tools in README.md if (required) with `./arkade get -o markdown`
- [x] I have updated the list of apps in README.md if (required) with `./arkade install --help`

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/alexellis/arkade/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: .github/workflows/check.yml [hunk 1]
```diff
@@ -0,0 +1,20 @@
+name: check-tools-upstream
+
+on:
+  workflow_dispatch:
+  schedule:
+    - cron: '0 * * * *'
```
Comment (alexellis) [inline .github/workflows/check.yml:6]: Add a comment in-line to show how frequently this is expected to run for anyone landing here who hasn't memorised cron expressions.

File: pkg/ci/main.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+	"log"
+	"net/http"
+	"sort"
+	"text/template"
+	"time"
+
+	"github.com/alexellis/arkade/pkg/get"
+)
+
+func URLExists(name, url, version string) error {
```
Comment (alexellis) [inline pkg/ci/main.go:15]: There's no need to make a new HTTP client every time this is called in a loop

File: pkg/ci/main.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+	}
+
+	if res.Body != nil {
+		defer res.Body.Close()
+	}
+
+	if res.StatusCode != http.StatusOK {
+		return fmt.Errorf("tool %s with version %s not found", name, version)
+	}
+	return nil
```
Comment (alexellis) [inline pkg/ci/main.go:37]: I think you could also check that the content-length returned is not 0?

File: pkg/ci/main.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+	"net/http"
+	"sort"
+	"text/template"
+	"time"
+
+	"github.com/alexellis/arkade/pkg/get"
+)
+
+func URLExists(name, url, version string) error {
+	fmt.Println("--------------->>>>>>>>>>>>>>>>>>>>>>>>>", name)
```
Comment (alexellis) [inline pkg/ci/main.go:16]: Let's print in the loop, not in the utility method.

File: pkg/ci/main.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+	timeout := time.Second * 10
```
Comment (alexellis) [inline pkg/ci/main.go:17]: Why did you pick 10 seconds for a `HEAD`? It should be instant.

File: pkg/ci/main.go [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+		defer res.Body.Close()
+	}
+
+	if res.StatusCode != http.StatusOK {
+		return fmt.Errorf("tool %s with version %s not found", name, version)
+	}
+	return nil
+}
+
+func main() {
```
Comment (alexellis) [inline pkg/ci/main.go:40]: Why did you pick a main and not Go's testing framework for this?

**Review by alexellis — COMMENTED** [review]
Overall I think this will be very useful. Added some comments, please review each and reply.

File: .github/workflows/check.yml (hunk 1 above)
Comment (yankeexe) [reply .github/workflows/check.yml:6]: okay, will update, how often do we want to run it? Once every day at 9 AM UTC?

File: pkg/ci/main.go (hunk 5 above)
Comment (yankeexe) [reply pkg/ci/main.go:17]: Testing on my local machine failed with 5sec time used in rest of the code base. Will revert it back to 5 sec.

File: pkg/ci/main.go (hunk 6 above)
Comment (yankeexe) [reply pkg/ci/main.go:40]: @alexellis  I don't understand what you mean?

File: pkg/ci/main.go (hunk 3 above)
Comment (yankeexe) [reply pkg/ci/main.go:37]: You mean content length == 0? As this is to check if no resource is returned from the request.

File: pkg/ci/main.go (hunk 6 above)
Comment (alexellis) [reply pkg/ci/main.go:36]: Could you convert the main.go into a unit test file please?

Then use t.Testing etc.

File: .github/workflows/check.yml (hunk 1 above)
Comment (alexellis) [reply .github/workflows/check.yml:6]: mid-night daily?

```
0 0 * * *
```

**Comment by alexellis** [comment]
Hi @yankeexe, it looks like you've been too busy to work on the comments, so to make things simpler and to get us covered, I've gone ahead and integrated this into arkade using the approach I mentioned above.

You have a credit for the suggestion in my commit message, but it is different code and integrated directly into the Tool struct.

https://github.com/alexellis/arkade/pull/636

Alex

**Comment by yankeexe** [comment]
@alexellis great addition! Will work on other stuffs, let me know if I can look into anything else.

**Pull request closed without merging**